### PR TITLE
Dagger

### PR DESCRIPTION
**Pull request by IamHaddad**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -44,4 +44,9 @@ dependencies {
     
     implementation constraintLayout
     implementation circularImageView
+
+    //Dagger2
+    implementation 'com.google.dagger:dagger:2.11'
```
Comment (helloseyedjafari) [inline app/build.gradle]: lotfan az file `dependencies.gradle` estefade konid

File: app/src/main/java/com/worldsnas/starplayer/MainApplication.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.worldsnas.starplayer
+
+import android.app.Application
+import com.worldsnas.starplayer.di.*
+
+class MainApplication : Application() {
+    companion object {
+        @JvmStatic
+        lateinit var mainApplicationComponent: MainApplicationComponent
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/MainApplication.kt]: khaheshan ta jayi ke mishe az `static` dori konid
toye `application` niazi nist static 
harja niaz bashe ba `context.aplicationContext` mishe gereftesh va rosh function call kard

File: app/src/main/java/com/worldsnas/starplayer/MainApplication.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        lateinit var playerFragmentComponent: PlayerFragmentComponent
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        mainApplicationComponent =
+            DaggerMainApplicationComponent.builder().androidmodule(Androidmodule(this)).build()
+        playerFragmentComponent =
+            DaggerPlayerFragmentComponent.builder().androidmodule(Androidmodule(this)).build()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/MainApplication.kt]: in mage baraye fragment nist?
chera inja dare sakhte mishe?

File: app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.worldsnas.starplayer.di
+
+import android.app.Application
+import android.content.Context
+import dagger.Module
+
+@Module
+class Androidmodule (private val application: Application){
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt]: lotfan az pass dadane maghadir be module ha dori konid
mitonid bejash az `@BindsInstance` estefade konid

File: app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+    fun provideApplicationContext():Context = application
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt]: bejaye function hayi ke faghat cast mikonan mitonin az `@Binds` estefade konin va `abstract module class`

File: app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.worldsnas.starplayer.di
+
+import android.app.Application
+import com.worldsnas.starplayer.MainActivity
+import dagger.Component
+
+@Component(modules = [Androidmodule::class])
+interface MainApplicationComponent {
+    fun inject (context: MainActivity)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt]: in component baraye kole application e
chera bayad `MainActivity` ro to in scope inject konim?
`MainActivity` age niaz dare mitone inject e khodesho dashte bashe

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.worldsnas.starplayer.di
+
+import com.worldsnas.starplayer.view.PlayerFragment
+import dagger.Component
+
+@Component (modules = [Androidmodule::class])
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: benazaram bad nist ye dependency be `ApplicationComponent` bedim
key dependency hayi ke to scope e application niaz hastan va provide mishan ro azash begirim

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 8]
```diff
@@ -21,4 +22,9 @@ class PlayerFragment : Fragment() {
         return inflater.inflate(R.layout.fragment_player, container, false)
     }
 
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        MainApplication.playerFragmentComponent.inject(this)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: behtare haminja sakhte beshe component

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
Good start
now let's refactor!!

File: app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt (hunk 6 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt]: یعنی یه کامپوننت برای اکتیویتی درست کنم؟

Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt]: age niaze bale

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt (hunk 7 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: اینو متوجه نشدم، میشه بیشتر توضیح بدید؟

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+    @Component.Builder
+    interface Builder {
+
+        fun build(): PlayerFragmentComponent
+
+        fun mainApplicationComponent(appComponent: AppComponent): Builder
+
+        @BindsInstance
+        fun playerFragment(playerFragment: PlayerFragment): Builder
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: niaz darim vaghean be fragment dakhele graph?

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 10]
```diff
@@ -21,4 +24,10 @@ class PlayerFragment : Fragment() {
         return inflater.inflate(R.layout.fragment_player, container, false)
     }
 
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        playerFragmentComponent = DaggerPlayerFragmentComponent.builder().build()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: parameter haro pass nadadin
crash mikone

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
much better 
good job

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt (hunk 9 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: اصولا مگه نباید به appComponent 
وصل باشه؟

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt (hunk 7 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: component ha mitonan be ham dependency dashte bashan

mamolan to android hame ye feature component ha be `AppComponent` niaz daran
manam goftam bad nist ma ham az hamin aval in dependency ro dashte bashim

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt (hunk 9 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: soaletono nafahmidam
man roye `PlayerFragment` comment gozashtam

Comment (neptu39701) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: منظور شما اینه که فرگمنت رو نیارم توی گراف appComponent 
درسته؟

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 10 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: یعنی appComponent?
رو اینطوری بهش بدم ؟
https://github.com/Shayan-Sakhaei/TasteV2/blob/70c48e3115aedd4cb7e5232a0b63693cca19d69b/app/src/main/java/ir/apptaste/android/view/fragment/DetailFragment.kt#L37

File: app/src/main/java/com/worldsnas/starplayer/di/AppComponent.kt [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package com.worldsnas.starplayer.di
+
+import android.app.Application
+import dagger.BindsInstance
+import dagger.Component
+
+@Component
+interface AppComponent {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/AppComponent.kt:10]: behtare scope e `Singleton` ro be appComponent bedim

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 10 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: injori ham mitonid

File: app/src/main/java/com/worldsnas/starplayer/MainActivity.kt [hunk 12]
```diff
@@ -2,11 +2,20 @@ package com.worldsnas.starplayer
...
 
 class MainActivity : AppCompatActivity() {
 
+    private lateinit var binding: ActivityMainBinding
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        setContentView(R.layout.activity_main)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        val view = binding.root
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/MainActivity.kt]: fek nemikonam be in `view` niazi bashe

File: app/src/main/java/com/worldsnas/starplayer/MainActivity.kt [hunk 13]
```diff
@@ -2,11 +2,20 @@ package com.worldsnas.starplayer
...
 import android.os.Bundle
+import androidx.annotation.IdRes
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentTransaction
+import com.worldsnas.starplayer.databinding.ActivityMainBinding
+import com.worldsnas.starplayer.view.musics_list.MusicsListFragment
 
 class MainActivity : AppCompatActivity() {
 
+    private lateinit var binding: ActivityMainBinding
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/MainActivity.kt]: nemishe `val` tarif kard?
fek konam to activity beshe

File: app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt (hunk 4 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt]: in file hamchenan niaze?

File: app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt (hunk 5 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/Androidmodule.kt]: inam niaze hanoz?

File: app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt (hunk 6 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/MainApplicationComponent.kt]: in hanoz niaze?

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
+package com.worldsnas.starplayer.di
+
+import com.worldsnas.starplayer.view.PlayerFragment
+import dagger.BindsInstance
+import dagger.Component
+@PlayerFragmentScope
+@Component(dependencies = [AppComponent::class])
+interface PlayerFragmentComponent {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentComponent.kt]: ahsant

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package com.worldsnas.starplayer.di
+
+import javax.inject.Scope
+
+@Scope
+@Retention(AnnotationRetention.RUNTIME)
+annotation class PlayerFragmentScopegit
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt]: esme behtar nemishe?!!!

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
good job

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt (hunk 15 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt]: on git ro havasam naboode bbakhshid, PFScope khobe?

File: app/src/main/java/com/worldsnas/starplayer/MainActivity.kt (hunk 13 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/MainActivity.kt]: توی raywenderlich همینو زده 
اگه بخوام جایگزین کنم، باید با nullable 
ها بنویسم یا 
با lazy

File: app/src/main/java/com/worldsnas/starplayer/MainActivity.kt (hunk 12 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/MainActivity.kt]: اینا رو من ننوشتم از dev گرفتم
ویرایش کنم؟

Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/MainActivity.kt]: bi zahmat!!

File: app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt (hunk 15 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt]: ScreenScope
ya 
FragmentScope
ya 
FeatureScope

ina chetoran?

Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/di/PlayerFragmentScope.kt]: FragmentScope khobe

**Review by helloseyedjafari — APPROVED** [review]
awesome

**Pull request merged**